src/spinalcordtoolbox/data/dataset/__init__.py

```diff
@@ -0,0 +1,2 @@
+# empty __init__.py to enable importlib.resources
+# see < TODO >
```

Comment:
There's another terrible python wart I ran into when setting this up that I don't know how or where to explain.


The basic issue is, `sys.path` behaves like $PATH or like overlayfs: it implies a virtual filesystem made of the union of all the folders listed in it. But python only pretends to support this and needs lots of workarounds for it. Recently, https://www.python.org/dev/peps/pep-0420/ dropped the necessity of `__init__.py` to create a package, in order to make it easier for this virtual union filesystem thing to work out; otherwise, every package that did what I did here, share a folder (`spinalcordtoolbox/data`), would be have to ship a `__init__.py` and would therefore be fighting with each other. So, A+ to that; but for some reason in the process they never applied the same thinking to non-python files. There's even this comment: https://github.com/python/cpython/blob/30e507dff465a31901d87df791a2bac40dc88530/Lib/importlib/resources.py#L243-L247

> By definition, namespace packages cannot have resources

So what this means is this terrible confusing situation where:

- if you `pip install` everything, everything ends up in the same physical directories and things make sense
- if you `pip install -e` one or more of the packages, then you're in this virtual situation where the filesystem you see via `import` and `importlib` is not the same as on disk but you can access python code but you can't access the others.

BUT they literally just fixed this a month ago: https://github.com/python/cpython/pull/24670. Praise the python maintainers. So maybe that link is what can go in the TODO.

Reply:
This tiny TODO took me like 4 hours to figure out. It definitely is where the bulk of my time working on this went and I don't know how to explain what I was doing because it's such a corner case.

And it has implications for the rest of sct. It means we need to delete all the __init__.pys from it. I checked and mostly they're empty anyway, but

* https://github.com/neuropoly/spinalcordtoolbox/blob/91e9764c7bce79972e445bb61c247ec2f00e49d2/spinalcordtoolbox/__init__.py
* https://github.com/neuropoly/spinalcordtoolbox/blob/91e9764c7bce79972e445bb61c247ec2f00e49d2/spinalcordtoolbox/utils/__init__.py

aren't totally.

But (until python3.10 I guess, when we finally get https://github.com/python/cpython/pull/24670) we'll have to keep these empty `__init__.py`s *in the data folders* but not in the *code* folders. Seems pretty fragile.

Or it looks like https://github.com/python/importlib_resources/ is still being maintained as a backport; I thought it was just for python3.6 but maybe it even covers newer versions?

Reply:
> And it has implications for the rest of sct. It means we need to delete all the **init**.pys from it. I checked and mostly they're empty anyway, but

Is this true? PEP420 [only applies to namespace packages](https://www.python.org/dev/peps/pep-0420/#specification). 

> Regular packages will continue to have an __init__.py and will reside in a single directory.
>
> Namespace packages cannot contain an __init__.py. 

Namespace packages are [defined as follows](https://packaging.python.org/guides/packaging-namespace-packages/):

> Namespace packages allow you to split the sub-packages and modules within a single package across multiple, separate distribution packages (referred to as distributions in this document to avoid ambiguity).

I don't think this describes the subpackages inside `spinalcordtoolbox`, does it? So, we should be able to leave the `__init__,py` files alone.

Reply:
I almost want to throw out `importlib.resources` entirely and just go back to using `__file__`. If we put:

```
# __init__.py

def contents():
    dir = os.path.dirname(__file__)
    return [os.path.join(dir, file) for file in os.listdir(dir)]
```

in every __init__.py we would have the same results for both

```
import spinalcordtoolbox.data.something

importlib.resources.contents(spinalcordtoolbox.data.something)
spinalcordtoolbox.data.something.contents()
```

 . It would mean we wouldn't be able to use virtual namespaces, so we couldn't make a multi-part `spinalcordtoolbox-data-something` package like python3.10 apparently can, but why would we want to?

...maybe we would want to do that to deal with splitting up data to fit online? Like a multi-part .rar or  a replacement for `ghsplit`?

Reply:
> import spinalcordtoolbox.data.something
> 
> importlib.resources.contents(spinalcordtoolbox.data.something)
> spinalcordtoolbox.data.something.contents()

This does feel a lot more readable. I was worried about `importlib` not being very self-evident.

Looking at the current usage, though (e.g. [`sct_deepseg`](https://github.com/neuropoly/spinalcordtoolbox/blob/master/spinalcordtoolbox/deepseg/models.py)) I feel like exposing the base directory might line up more with how it parses for files.

Reply:
Okay actually I think this provides a solution:

* https://github.com/python/importlib_resources/issues/68 -> https://github.com/python/importlib_resources/pull/196 fixed this in `importlib_resources`, the backport
* https://bugs.python.org/issue42129 -> https://github.com/python/cpython/pull/24670 fixes this in `importlib.resources`

So we can just:

```
install_requires=[
  ...
  importlib_resources>=4
  ...
],
```

and

```
import importlib_resources as resources
# from importlib import resources # uncomment when https://github.com/python/cpython/pull/24670 is available in python3.10
```

Reply:
I still kind of like just using `__file__`. But `resources` is the Way Of The Future so maybe we should go with that.